R/R/prophet.R

```diff
@@ -1274,37 +1268,203 @@ fit.prophet <- function(m, df, ...) {
 #' forecast <- predict(m, future)
 #' plot(m, forecast)
 #' }
-#'
 #' @export
-predict.prophet <- function(object, df = NULL, ...) {
+predict.prophet <- function(m, df = NULL, ...) {
+  
   if (is.null(df)) {
-    df <- object$history
+    df <- m$history
   } else {
     if (nrow(df) == 0) {
       stop("Dataframe has no rows.")
     }
-    out <- setup_dataframe(object, df)
+    out <- setup_dataframe(m, df)
     df <- out$df
+    m <- out$m
   }
-
-  df$trend <- predict_trend(object, df)
-  seasonal.components <- predict_seasonal_components(object, df)
-  intervals <- predict_uncertainty(object, df)
-
-  # Drop columns except ds, cap, floor, and trend
-  cols <- c('ds', 'trend')
-  if ('cap' %in% colnames(df)) {
-    cols <- c(cols, 'cap')
+  
+  # Construct input to stan
+  out.fit <- make_all_seasonality_features(m, m$history)
+  out.predict <- make_all_seasonality_features(m, df)
+  
+  dat <- data_for_stan(
+    m, 
+    df = df, 
+    fit.seasonal.features = out.fit$seasonal.features,
+    predict.seasonal.features = out.predict$seasonal.features, 
+    component.cols = m$train.component.cols, 
+    prior.scales = out.fit$prior.scales, 
+    type = "predict"
+  )
+  
+  # Run stan
+  if (exists(".prophet.stan.model")) {
+    model <- .prophet.stan.model
+  } else {
+    model <- get_prophet_stan_model()
   }
-  if (object$logistic.floor) {
-    cols <- c(cols, 'floor')
+  
+  stan_init <- 
+    list(
+      k = m$params$k,
+      m = m$params$m,
+      delta = as.numeric(m$params$delta),
+      beta = as.numeric(m$params$beta),
+      sigma_obs = m$params$sigma_obs
+    )
+  
+  if (m$mcmc.samples > 0) {
+    args <- list(
+      object = model,
+      data = dat,
+      init = stan_init,
+      iter = m$mcmc.samples
+    )
+    args <- utils::modifyList(args, list(...))
+    stan.fit <- do.call(rstan::sampling, args)
+  } else {
+    args <- list(
+      object = model,
+      data = dat,
+      init = stan_init,
+      iter = 0,
+      as_vector = FALSE,
+      algorithm = 'Newton',
+      verbose = FALSE
+    )
+    args <- utils::modifyList(args, list(...))
```

Comment:
Actually let's don't do this. In `fit`, we accept any args and pass them along to stan to give more control over the fitting (especially if you're doing MCMC). Here in `predict`, `predict.prophet` accepts `...` args because it is required in order to implement `predict`, but we don't want to pass them along to Stan. Since we're not actually doing any fitting, there's no need to give control here.